### ISSUE TITLE
Improve reusability utils

### ISSUE THREAD

**Issue by japgolly** [japgolly/scalajs-react#302]
`Reusability` itself is awesome.

`ReusableFn` is pretty good.

`ReusableVar` is good on its own but something about it's relationship to `ExternalVar` seems off. If nothing else the names should be sync'd a bit better.

`ReusableVal` and `ReusableVal2` feel like failed experiments. Precisely why will need reflection.

Also here's another idea that makes ReusableFn final values reusable:

``` scala
case class ReusableCoState[S, A](fn: S ~=> A, s: S) {
  def value: A = fn(s)
}
object ReusableCoState {
  // TODO Remove Reusability bound on S, get from fn
  implicit def reusability[S: Reusability, A]: Reusability[ReusableCoState[S, A]] =
    Reusability.fn[ReusableCoState[S, A]]((x, y) =>
      (x.fn ~=~ y.fn) && (x.s ~=~ y.s)
    )
}
```

**Comment by japgolly** [comment]
Another thing is that the choice between reusability (eg. `ReusableVar`) and non-reusability (eg. `ExternalVar`) is made at the component level then affects types of entire call graph.

Better way of thinking of would be
1. If component support reusability, declare it.
2. Call-sites decide whether they want reusability or not.

This would mean the reusability logic is provided by call-site rather than at component construction. If so, cases where reuse isn't desired `Reusability.never` should be used under-the-hood.

If reusability logic is provided downstream then distinct types like `{External,Reusable}Var` needn't exist; they just become one.

But am I not already getting this at the type-level rather than at the value-level?

Basically the difference between `A => B` and `A ~=> B` is what needs more attention.

**Comment by n3ziniuka5** [comment]
Also, how would you solve reusability when you want a reusable callback without parameters? For example you have a todo list and each item in the list can be deleted on click. The item would have props
`
case class Props(
  text: String,
  onClick: Callback // () ~=> Callback is invalid
)
`

Then on parent component you would like to pass in a function
`
val onTodoItemClick = ReusableFn((id: String) =>
  // delete item from state
)
...
//create component
TodoItem("item 1", onTodoItemClick("id1"))
TodoItem("item 2", onTodoItemClick("id2"))
`

**Comment by n3ziniuka5** [comment]
Also,
```scala
val fn = ReusableFn((arg1: String, arg2: String) =>
  ...
)
```
This produces `ReusableFn[String, ReusableFn[String, Callback]]` or `String ~=> (String ~=> Callback)` which is very counter-intuitive. Maybe have `ReusableFn[String, String, Callback]` and `(String, String) ~=> Callback`?

**Comment by japgolly** [comment]
Fixed on `topic/neo` branch via aae160f9a4d6d0feca20e1ba0fc10d10185c7cd1.

Docs haven't been updated yet.

Basic summary:
* `Reusability[A]`: Encodes ∀a∈A.
* `Reusable[A]`: Encodes ∃a∈A. Replaces `ReusableVal` and `ReusableVal2`.
* `ReusableFn[A, B]`: Remains as is (although it could be represented by `Reusable[A => B]` minus all the composition methods).
* `ReusableFnA[A, B]`: Gone. Failed experiment and/or rare usecase. `Reusable[B]` suffices.
* `StateSnapshot[A]`: Replaces `ExternalVar` and `ReusableVar`. Reuse optional. Switching it on/off no longer painful.

@n3ziniuka5 For your case there's no reusability logic that holds ∀ `Callback`s (except reference comparison which is nearly always too weak), but you can now use the ∃ version instead, i.e. instead of passing around `Callback`s where you want compile-time proven reusability, pass around `Reusable[Callback]` instead and have the call-site specify the reuse logic that applies to its specific scope and use-case.
